go/arrow/util/protobuf_reflect_test.go

```diff
@@ -0,0 +1,180 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"github.com/apache/arrow/go/v16/arrow/array"
+	"github.com/apache/arrow/go/v16/arrow/memory"
+	"github.com/apache/arrow/go/v16/arrow/util/util_message"
+	"github.com/huandu/xstrings"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/types/known/anypb"
+	"strings"
+	"testing"
+)
+
+func SetupTest() util_message.AllTheTypes {
+	msg := util_message.ExampleMessage{
+		Field1: "Example",
+	}
+
+	anyMsg, _ := anypb.New(&msg)
+
+	return util_message.AllTheTypes{
+		String_:     "Hello",
+		Int32:       10,
+		Int64:       100,
+		Sint32:      -10,
+		Sin64:       -100,
+		Uint32:      10,
+		Uint64:      100,
+		Fixed32:     10,
+		Fixed64:     1000,
+		Sfixed32:    10,
+		Bool:        false,
+		Bytes:       []byte("Hello, world!"),
+		Double:      1.1,
+		Enum:        util_message.AllTheTypes_OPTION_0,
+		Message:     &msg,
+		Oneof:       &util_message.AllTheTypes_Oneofstring{Oneofstring: "World"},
+		Any:         anyMsg,
+		SimpleMap:   map[int32]string{99: "Hello", 100: "World"},
+		ComplexMap:  map[string]*util_message.ExampleMessage{"complex": &msg},
+		SimpleList:  []string{"Hello", "World"},
+		ComplexList: []*util_message.ExampleMessage{&msg},
+	}
+}
+
+func TestGetSchema(t *testing.T) {
+	msg := SetupTest()
+
+	pr := NewProtobufStructReflection(&msg)
+	got := pr.GetSchema().String()
+	want := `schema:
+  fields: 22
+    - string: type=utf8, nullable
+    - int32: type=int32, nullable
+    - int64: type=int64, nullable
+    - sint32: type=int32, nullable
+    - sin64: type=int64, nullable
+    - uint32: type=uint32, nullable
+    - uint64: type=uint64, nullable
+    - fixed32: type=uint32, nullable
+    - fixed64: type=uint64, nullable
+    - sfixed32: type=int32, nullable
+    - bool: type=bool, nullable
+    - bytes: type=binary, nullable
+    - double: type=float64, nullable
+    - enum: type=dictionary<values=utf8, indices=int32, ordered=false>, nullable
+    - message: type=struct<field1: utf8>, nullable
+    - oneofstring: type=utf8, nullable
+    - oneofmessage: type=struct<field1: utf8>, nullable
```

Comment:
shouldn't we use a Union for `oneof`s?

Reply:
Do you have a preference for dense or sparse? I don't think I do.

Reply:
I don't have a preference. I'd probably go with dense unless someone asks differently. Perhaps make it configurable somehow?

Reply:
I can't find a nice way to implement this. 
As I loop through the fields of the struct all of the options in the one of are their own field. 
I would need to gather these after, and collect them into a dense union. 

Is it acceptable to leave this as extendable? Otherwise I need some help.

Reply:
yea, it's fine to leave it as extendable for now

Reply:
having confessed I don't know what I'm doing I had an idea, I'll keep at it, but good to know

Reply:
can work with DenseUnion now 👍 , I'll do some tidying up tomorrow the super field duplicates the protobufFieldReflection.